### PR TITLE
Fix disabled items preventing research completion

### LINKED ISSUE #3782: sf stats counting disabled items in the list of unlocked items, making us unable to achieve 100%
### ❗ Checklist

- [X] I am using the official english version of Slimefun and did not modify the jar.
- [X] I am using an up to date "DEV" (not "RC") version of Slimefun.
- [X] I am aware that issues related to Slimefun addons need to be reported on their bug trackers and not here.
- [X] I searched for similar open issues and could not find an existing bug report on this.

### 📍 Description

sf stats has a place where it shows total researches, and its counting the disabled ones as  a part of it

### 📑 Reproduction Steps

disable a item in sf and then unlock everything else, then check sf stats but you wont have 100% researches

### 💡 Expected Behavior

sf stats should not include disabled items into calculating % of  research done by player

### 📷 Screenshots / Videos

image
everything has been unlocked but still due to disabled items player is unable to reach 100% 

### 📜 Server Log

_No response_

### 📂 `/error-reports/` folder

_No response_

### 💻 Server Software

Paper

### 🎮 Minecraft Version

1.19.x

### ⭐ Slimefun version

Sf version 1043 

### 🧭 Other plugins

this bug is not related to a plugin.

### PR DESCRIPTION
**Pull request by test137E29B**
## Description

When a SlimefunItem bound to a Research is disabled, the Research is not disabled. This means that the Research becomes locked permanently and the user can never achieve 100% Research unlock completion. If a SlimefunItem is globally disabled, it should not be bindable to Research since it's never usable. If a Research has no bound SlimefunItems (because they're all globall disabled), the research should act as not enabled also. Disabled researches should be filtered out too from the Progress stat.

## Proposed changes

Prevent disabled items being added to Research, filter out Researches with no Items bound when requested. This could also be triggered some other place that's more efficient than every time research is requested. Please suggest another place to perform this action if a better place exists.

## Related Issues (if applicable)

Resolves #3782 

## Checklist

- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.19.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/researches/Research.java [hunk 1]
```diff
@@ -84,12 +84,13 @@ public Research(@Nonnull NamespacedKey key, int id, @Nonnull String defaultName,
     /**
      * This method returns whether this {@link Research} is enabled.
      * {@code false} can mean that this particular {@link Research} was disabled or that
-     * researches altogether have been disabled.
+     * researches altogether have been disabled, or that this {@link Research} as no bound
```
Comment (test137E29B) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/researches/Research.java:87]:
```suggestion
     * researches altogether have been disabled, or that this {@link Research} has no bound
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/player/PlayerProfile.java [hunk 2]
```diff
@@ -12,6 +12,7 @@
 import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Level;
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/player/PlayerProfile.java:14]: Unused import

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

**Comment by JustAHuman-xD** [comment]
For an example of this fix see: https://github.com/JustAHuman-xD/Slimefun4/commit/180731553df56eba65f3126ccc8d644bed283502

**Comment by JustAHuman-xD** [comment]
For an example of this fix see: https://github.com/JustAHuman-xD/Slimefun4/commit/180731553df56eba65f3126ccc8d644bed283502

**Comment by test137E29B** [comment]
> For an example of this fix see: https://github.com/JustAHuman-xD/Slimefun4/commit/180731553df56eba65f3126ccc8d644bed283502

But if you bind 2x researches to the same item, it unbinds the item from one of the researches (the first), thus leaving a research with NO items. This research is then NOT unlockable and again causes the same problem where you can't 100% research.

Researches with no items should be treated as not enabled because they are incomplete and cannot work correctly. This was conciously changed globally to prevent anywhere research is used from taking into account any research that is not set up correctly.

**Comment by JustAHuman-xD** [comment]
Idk it still feels like an iffy change but others can give their opinion on it. I think that way because people could make dummy/marker researches with no items in it intentionally, or disabled items intentionally tho a reason for that I can't think of off the top of my head. And the change you made to getResearches is definitely an iffy change. I would suggest instead adding a method called getEnabledResearches. As getResearches implies all researches not just enabled ones.

**Comment by JustAHuman-xD** [comment]
TLDR changing the behavior of old and crucial API is always scary

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/SlimefunRegistry.java [hunk 3]
```diff
@@ -200,7 +201,9 @@ public List<SlimefunItem> getEnabledSlimefunItems() {
      */
     @Nonnull
     public List<Research> getResearches() {
```
Comment (Sefiraat) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/SlimefunRegistry.java:203]: When requesting a Set of all researches, one would expect to get the whole list, regardless of the researches current state.

If you want a list of enabled researches only, then a new method to return this should be made.

Comment (test137E29B) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/SlimefunRegistry.java:203]: Added a new method for this, should have done it this way to begin with - also fixed other usages and added specific placeholder for total researches enabled as `researches_total_researches_enabled`

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/items/SlimefunItem.java [hunk 4]
```diff
@@ -649,6 +649,11 @@ private void checkForDeprecations(@Nullable Class<?> c) {
      *            The new {@link Research} for this {@link SlimefunItem}, or null
      */
     public void setResearch(@Nullable Research research) {
+        if (this.isDisabled()) {
```
Comment (test137E29B) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/items/SlimefunItem.java:652]: Should we still allow disabled items to be bound to researches?

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/researches/Research.java [hunk 5]
```diff
@@ -84,12 +84,13 @@ public Research(@Nonnull NamespacedKey key, int id, @Nonnull String defaultName,
...
      * {@code false} can mean that this particular {@link Research} was disabled or that
-     * researches altogether have been disabled.
+     * researches altogether have been disabled, or that this {@link Research} has no bound
+     * {@link SlimefunItem} and is therefore not-unlockable.
      * 
      * @return Whether this {@link Research} is enabled or not
      */
     public boolean isEnabled() {
-        return Slimefun.getRegistry().isResearchingEnabled() && enabled;
+        return Slimefun.getRegistry().isResearchingEnabled() && enabled && items.size() > 0;
```
Comment (test137E29B) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/researches/Research.java:93]:
```suggestion
        return Slimefun.getRegistry().isResearchingEnabled() && enabled;
```

**Comment by test137E29B** [comment]
**NB** adds new Placeholder `researches_total_researches_enabled`

File: src/main/java/io/github/thebusybiscuit/slimefun4/api/items/SlimefunItem.java (hunk 4 above)
Comment (test137E29B) [inline src/main/java/io/github/thebusybiscuit/slimefun4/api/items/SlimefunItem.java:652]: If this check is not here, we'd have to check every affected item when getting enabled researches.

I don't think this is a good solution anymore though tbh

**Comment by JustAHuman-xD** [comment]
Superceded by #3917
